MyLazy/MyLazy/ILazy.cs

```diff
@@ -0,0 +1,13 @@
+﻿namespace MyLazy
+{
+    /// <summary>
+    /// Interface for Lazy computation
+    /// </summary>
+    public interface ILazy<T>
```

Comment:
```suggestion
    public interface ILazy<out T>
```
Вот хорошее объяснение вариантности: http://tomasp.net/blog/variance-explained.aspx/